internal/app/machined/pkg/system/runner/process/process.go

```diff
@@ -158,37 +251,60 @@ func (p *processRunner) build() (commandWrapper, error) {
 			return commandWrapper{}, err
 		}
 
-		cmd.Stderr = stderr
+		wrapper.stderr = stderr.Fd()
 
 		afterStartFuncs = append(afterStartFuncs, func() {
 			stderr.Close() //nolint:errcheck
 		})
 	} else {
-		cmd.Stderr = writer
+		wrapper.stderr = pw.Fd()
 	}
 
-	ctty, cttySet := p.opts.Ctty.Get()
-	if cttySet {
-		cmd.SysProcAttr = &syscall.SysProcAttr{
-			Setsid:  true,
-			Setctty: true,
-			Ctty:    ctty,
+	closeWriter = false
+
+	wrapper.launcher = launcher
+	wrapper.afterStart = func() {
+		for _, f := range afterStartFuncs {
+			f()
 		}
 	}
+	wrapper.afterTermination = closeLogging
+	wrapper.ctty = p.opts.Ctty
 
-	closeWriter = false
+	return wrapper, nil
+}
 
-	return commandWrapper{
-		cmd: cmd,
-		afterStart: func() {
-			for _, f := range afterStartFuncs {
-				f()
-			}
-		},
-		afterTermination: func() error {
-			return w.Close()
-		},
-	}, nil
+// Apply cgroup and OOM score after the process is launched.
+func applyProperties(p *processRunner, pid int) error {
+	path := cgroup.Path(p.opts.CgroupPath)
+
+	if cgroups.Mode() == cgroups.Unified {
+		cgv2, err := cgroup2.Load(path)
+		if err != nil {
+			return fmt.Errorf("failed to load cgroup %s: %w", path, err)
+		}
+
+		if err := cgv2.AddProc(uint64(pid)); err != nil {
```

Comment:
can we move to cgroupFD?

Reply:
Don't we also use this launch method when running in a container (thus not sure we have clone3 available)? Cgroup in clone3 is actually only available since kernel 5.7

Reply:
good point, let's skip it for now

Reply:
can we optionally have two paths? like auto-detect?

Reply:
Unsure whether that's going to be beneficial, but it's possible for sure

Reply:
clone3 into cgroup is only possible with cgv2 ([ref](https://github.com/systemd/systemd/blob/e8fb0643c1bea626d5f5e880c3338f32705fd46d/src/basic/process-util.c#L2043)). Will implement clone3 for the case we are running native with cgv2.

Reply:
we'll always be running with cgroupv2, cgroupv1 is also not fully removed from systemd, so that's the general directions from all distros

Reply:
We still support cgroupsv1 both for native launch, and we'll have to support in container mode.